src/net/db.rs

```diff
@@ -363,6 +389,28 @@ const PEERDB_INITIAL_SCHEMA: &'static [&'static str] = &[
 const PEERDB_INDEXES: &'static [&'static str] =
     &["CREATE INDEX IF NOT EXISTS peer_address_index ON frontier(network_id,addrbytes,port);"];
 
+const PEERDB_SCHEMA_2: &'static [&'static str] = &[
+    r#"
+    CREATE TABLE stackerdb_peers(
+        smart_contract_id TEXT NOT NULL,
+        peer_slot INTEGER NOT NULL,
```

Comment:
I might be completely misunderstanding the question but here are my thoughts.

- slot seems to be calculated like this `hash(addr + ip)`
- according to [this](https://kevingal.com/apps/collision.html) the change for collision if there are 1000 peers is ~2.7089e-14
- is it possible for a different peer to get the same `(addr, port)` pair over time?

Reply:
Some background.

The Stacks node maintains a table called `frontier` which stores a record of every other peer it has ever contacted.  Maintaining this "phonebook" helps the peer recover from periods of intermittent network unavailability, including node crashes -- once the peer's network becomes available, it already has a persistent set of neighbors it can sample in order to re-establish itself as a node in the K-regular random peer graph.

Peers are discovered through the neighbor-walk state machine.  The neighbor-walk state machine itself is somewhat complex so I'll omit the details here, but the end result of the state-machine's execution is that it writes new rows to the `frontier` for newly-discovered neighbors.

Because peers can return any neighbor data they want, a malicious peer could take the opportunity to flood a victim's `frontier` table with garbage, leading to a DoS by resource exhaustion.  It can also report peer addresses that conflict with existing peers in the victim's `frontier` table.  To counter resource exhaustion, the the `frontier` table's size is necessarily size-bound -- in this case, up to 2**24 records may be inserted.

When inserting a peer into the frontier, each peer is hashed to eight different slots (rows).  If _at least one of them_ corresponds to an empty row in `frontier`, then the peer is inserted.  However, if _all eight_ slots are occupied, then the insert fails and the caller must first evict at least one colliding peer.

The neighbor-walk state machine is the only subsystem that ever inserts peers into the `frontier` table, and it handles the eviction case.  Specifically, it pings the peer to be evicted, and if that peer _does not respond_, then the newly-discovered peer is inserted.  Otherwise, the newly-discovered peer is dropped.  This approach is based on recommendations from [this USENIX paper](https://www.usenix.org/system/files/conference/usenixsecurity15/sec15-paper-heilman.pdf) on eclipse attacks on Bitcoin's p2p network.

---

The eviction strategy guarantees that `frontier.slot` values are unique.  A peer's row is not added if its slot ID is already present in the DB.  A peer's row can only be updated specifically in the case of peer eviction (in which case, the slot ID is preserved).

`peer_slot` is not a foreign key because `slot` cannot be a primary key.  Different peers can hash to the same slot IDs, but we would want to store both peers.

Reply:
Okay, but `slot` _is_ the primary key of the `frontier` table unless I'm really misreading something.

Reply:
`peer_slot` is now treated as a `FOREIGN KEY` to `frontier`